File: javascript/src/linklocator.js
function distance(x1, y1, x2, y2){
    return Math.sqrt(Math.pow((x1 - x2), 2) + Math.pow((y1 - y2), 2))
}

function power(reach, distance){
    if (distance >= reach) return 0 
    return Math.pow((reach - distance), 2)
}

function trimSearch(stations, device){
    function withinXReach(device) {
        return function(element) {
            return Math.abs(element[0] - device[0]) < element[2]
        }
    }
    
    function withinYReach(device) {
        return function(element) {
            return Math.abs(element[1] - device[1]) < element[2]
        }
    }

    stations = stations.filter(withinXReach(device))
    stations = stations.filter(withinYReach(device))

    return stations
}

function bestlinks(stations, devices){
    let bestlinks = {}
    console.log(stations)
    console.log(devices)
    for (const device of devices) {
        bestlinks[device] = {device : device}
        let trimedStations = trimSearch(stations, device)
        if (trimedStations.length == 0) continue
        for (const station of trimedStations) {
            let linkdistance = distance(device[0], device[1], station[0], station[1])
            let linkpower = power(station[2], linkdistance) 
            if (linkpower == 0) continue
            if ('power' in bestlinks[device]){
                if (linkpower < bestlinks[device]['power']) continue
            }
            bestlinks[device]['station'] =  station
            bestlinks[device]['distance'] =  linkdistance
            bestlinks[device]['power'] =  linkpower
        }        
    }
    return bestlinks
}

module.exports = {
    bestlinks,
    distance,
    power,
    trimSearch
}

File: python/src/linklocator.py
import math
def distance(x1, y1, x2, y2):
    return math.sqrt(math.pow((x1 - x2), 2) + math.pow((y1 - y2), 2))

def power(reach, distance):
    if (distance >= reach):
        return 0 
    return math.pow((reach - distance), 2)


def trimSearch(stations, device):
    def withinXReach(station, device) :
        return abs(station[0] - device[0]) < station[2]
    
    def withinYReach(station, device) :
        return abs(station[1] - device[1]) < station[2]      
    
    stations = filter(lambda station: withinXReach(station, device), stations)
    stations = filter(lambda station: withinYReach(station, device), stations)

    return stations


def bestlinks(stations, devices):
    bestlinks = {}
    for x in range(0, len(devices)):
        device = devices[x]        
        bestlinks[tuple(device)] = {'device' : device}
        trimedStations = trimSearch(stations, device)
        if (len(trimedStations) == 0):
             continue
        for y in range(0, len(trimedStations)):
            station = trimedStations[y]
            linkdistance = distance(device[0], device[1], station[0], station[1])
            linkpower = power(station[2], linkdistance) 
            if (linkpower == 0):
                continue
            if ('power' in bestlinks[tuple(device)]):
                if (linkpower < bestlinks[tuple(device)]['power']):
                    continue
            
            bestlinks[tuple(device)]['device'] =  device
            bestlinks[tuple(device)]['station'] =  station
            bestlinks[tuple(device)]['distance'] =  linkdistance
            bestlinks[tuple(device)]['power'] =  linkpower
                   
    return bestlinks